### DISCUSSION TITLE
Get a repository's commits along with changed patches and the url to changed files using GraphQL v4

### DISCUSSION THREAD

**Discussion by armsp** [community/community#24496] [Apps, API and Webhooks]
<p>I am able to get the list of <em>commits</em> (with fields like <em>commit messages</em>, <em>oid</em>, <em>commit url</em> etc) along with the number of changedFiles, made in a repository on the master branch.<br>
However I am not able to figure out how to get  <strong>any information about the changes themselves and the files that were changed.</strong></p>
<p>In v3 of the REST API, the information about the changes were contained in <em>files</em>-&gt;<em>patch</em>, and <em>files</em> -&gt; <em>raw_url</em> or <em>blob_url</em> gave info about the original file itself at that stage.</p>
<p>Q) In v4 of GitHub’s API using GraphQL how do I get the corresponding information?</p>
<p>This is the query I am stuck with right now (showing only 1 commit for brevity) -</p>
<pre><code>query {
  rateLimit{
    cost
    remaining
  }
  repository(owner: "elastic", name: "elasticsearch") {
    name
    defaultBranchRef {
      name
      target {
        ... on Commit {
          history(first:1){
            nodes{
              message
              changedFiles
              id
              oid
              treeUrl
              url
              tree{
                oid
              }
            }
            pageInfo{
              hasNextPage
              startCursor
              endCursor
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>Output:</p>
<pre><code>{
  "data": {
    "rateLimit": {
      "cost": 1,
      "remaining": 4999
    },
    "repository": {
      "name": "elasticsearch",
      "defaultBranchRef": {
        "name": "master",
        "target": {
          "history": {
            "nodes": [
              {
                "message": "Small corrections to HLRC doc for _termvectors (#35221)\n\nRelates to #33447",
                "changedFiles": 2,
                "id": "MDY6Q29tbWl0NTA3Nzc1OmEyYzIyYWQ3YWViMGY4ZDUxNDg2NzdkZDcyMjJhZDQzYWZlZTlhMTc=",
                "oid": "a2c22ad7aeb0f8d5148677dd7222ad43afee9a17",
                "treeUrl": "https://github.com/elastic/elasticsearch/tree/a2c22ad7aeb0f8d5148677dd7222ad43afee9a17",
                "url": "https://github.com/elastic/elasticsearch/commit/a2c22ad7aeb0f8d5148677dd7222ad43afee9a17",
                "tree": {
                  "oid": "4f5f11e0e55aeafc4677800959232726a2cd787c"
                }
              }
            ],
            "pageInfo": {
              "hasNextPage": true,
              "startCursor": "a2c22ad7aeb0f8d5148677dd7222ad43afee9a17 0",
              "endCursor": "a2c22ad7aeb0f8d5148677dd7222ad43afee9a17 0"
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p> I tried many ways to somehow extract information about the patches however I failed all the time. I have started learning about GraphQL just a couple of days ago so I don’t have any experience with it.  </p>
<p>I was thinking if there is a corresponding object for <em>patch</em>(that shows the changes in REST API v3) in GraphQL API then I might be able to reverse engineer my way through it, howver the docs have been very cryptic.</p>
<p>Any guidance would be much appreciated.</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/armsp">@armsp</a>, </p>
<p>Thank you for being here! We don’t yet support that in the GraphQL API. There has been a schema request for it. I’ll update this post when we get new info about a ship.</p>
<p>Best,</p>
<p>Andrea</p>

**Comment by platy** [comment]
<p>Hi Andrea,</p>
<p>should we still be using v3 for getting information about the changes in a commit? I think a comment on the <a href="https://developer.github.com/v4/object/commit/" rel="nofollow noopener">v4 Commit object documentation</a> to say that would be really helpful.</p>
<p>Thanks,</p>
<p>Mike</p>

**Comment by AndreaGriffiths11** [comment]
<p>You are correct <a class="mention" href="/platy">@platy</a> and thank you for the feedback I’ll pass it along to the documentation team. Thanks again!</p>

**Comment by jamieoneill** [comment]
<p>any news on this? i still dont see it documented</p>
<p>Thanks</p>

**Comment by bobbytreed** [comment]
<p>Still looking for this as well.</p>

**Comment by lots0logs** [comment]
<p><img src="https://user-images.githubusercontent.com/4675662/181085056-7ab0db73-7ce5-4953-aa6c-ca3b7e76a748.gif" alt="bump.gif" data-base62-sha1="4iXCaoMtqjf7YIzSaK68AJzDdb4" width="319" height="257"></p>

**Comment by mljrg** [comment]
<p>Hi Andrea,</p>
<p>can’t the technical team include the information about file changes, i.e., the paths of added, modified, deleted, and renamed files, in <a href="https://developer.github.com/v4/object/commit/" rel="nofollow noopener">GraphQL’s Commit</a> object?</p>
<p>This can save a lot of network resources when compared to getting the same information via the Rest API <a href="https://developer.github.com/v3/repos/commits/#get-a-single-commit" rel="nofollow noopener">Get a single commit</a> , which always sends patches over the network …</p>
<p>Thanks</p>

**Comment by wbuiejr** [comment]
<p>To advertise GraphQL <a href="https://developer.github.com/v4/#why-github-is-using-graphql" rel="nofollow noopener">claiming it offers greater flexibility</a>, and then not offer the flexibility in v3 does not give me warm fuzzy feelings.</p>
<p>Is it safe to assume that v3 will continue to be supported at least until everything I can do in v3 is possible in v4?</p>

**Comment by pconrad** [comment]
<p>Is it still the case that in v4 with GraphQL, that you cannot get a list of the files changed in a commit?</p>
<p>And if so, what’s the tl;dr on what you need to do using the v3 api to get around this limitation?</p>
<p>I am writing scripts to analyze commit behavior on student software engineering projects.   I need to get a list of the files changed so that we can exclude certain commits that only touch documentation and not code.</p>

**Comment by MaxAst** [comment]
<p>+1 on needing this functionality. I need to get a list of file extensions per commit</p>

**Comment by francisfuzz** [comment, marked as answer]
<aside class="quote no-group" data-username="mljrg" data-post="8" data-topic="13585">
<div class="title">
<div class="quote-controls"></div>
 mljrg:</div>
<blockquote>
<p>can’t the technical team include the information about file changes, i.e., the paths of added, modified, deleted, and renamed files, in <a href="https://developer.github.com/v4/object/commit/">GraphQL’s Commit </a> object?</p>
<p>This can save a lot of network resources when compared to getting the same information via the Rest API <a href="https://developer.github.com/v3/repos/commits/#get-a-single-commit">Get a single commit </a> , which always sends patches over the network …</p>
</blockquote>
</aside>
<aside class="quote no-group" data-username="MaxAst" data-post="11" data-topic="13585" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 MaxAst:</div>
<blockquote>
<p>+1 on needing this functionality. I need to get a list of file extensions per commit</p>
</blockquote>
</aside>
<p>:wave: <a class="mention" href="/mljrg">@mljrg</a>: These are great points to bring up and something our product team would be interested in hearing more about. The best place to share product feedback and feature requests is through <a href="https://support.github.com/contact/feedback">our official product feedback form</a>.</p>
<hr>
<aside class="quote no-group" data-username="wbuiejr" data-post="9" data-topic="13585" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 wbuiejr:</div>
<blockquote>
<p>To advertise GraphQL <a href="https://developer.github.com/v4/#why-github-is-using-graphql">claiming it offers greater flexibility </a>, and then not offer the flexibility in v3 does not give me warm fuzzy feelings.</p>
<p>Is it safe to assume that v3 will continue to be supported at least until everything I can do in v3 is possible in v4?</p>
</blockquote>
</aside>
<p>:wave: <a class="mention" href="/wbuiejr">@wbuiejr</a>: Thanks for bringing up that point. I can see how that wording can come across and there’s definitely an opportunity for us to review the current presentation; this is something that our product team is open to hearing about as well.</p>
<p>Having feature parity between the GitHub REST API and the GraphQL is something that the API team is aware of, though I personally can’t speak if or when there will be an initiative to move that forward.</p>
<p>If it helps, we recently announced the <a href="https://github.community/t/github-roadmap-faq/124978">GitHub public roadmap</a>. This is the one place where anyone can learn about what features our team is working on, what stage they’re in, and when they’re expected to be released.</p>
<p>At this stage, we’re going to continue to support both the GitHub REST API and the GitHub GraphQL API.</p>
<hr>
<aside class="quote no-group" data-username="pconrad" data-post="10" data-topic="13585">
<div class="title">
<div class="quote-controls"></div>
 pconrad:</div>
<blockquote>
<p>Is it still the case that in v4 with GraphQL, that you cannot get a list of the files changed in a commit?</p>
</blockquote>
</aside>
<p>:wave: <a class="mention" href="/pconrad">@pconrad</a>: The <a href="https://docs.github.com/en/graphql/reference/objects#commit"><code>Commit</code> object</a> doesn’t expose a field that lists the files changed. However, it still exposes a <code>changedFiles</code> field indicating the number of changed files.</p>
<aside class="quote no-group" data-username="pconrad" data-post="10" data-topic="13585">
<div class="title">
<div class="quote-controls"></div>
 pconrad:</div>
<blockquote>
<p>And if so, what’s the tl;dr on what you need to do using the v3 api to get around this limitation?</p>
</blockquote>
</aside>
<p>The REST API exposes an <a href="https://docs.github.com/en/rest/reference/repos#get-a-commit">endpoint for fetching a single commit</a>. The API responds with a JSON object with a <code>files</code> attribute listing all of the files changed. This endpoint also supports the <a href="https://developer.github.com/v3/media/#commits-commit-comparison-and-pull-requests">diff and patch</a> media types. Does this help?</p>
<aside class="quote no-group" data-username="pconrad" data-post="10" data-topic="13585">
<div class="title">
<div class="quote-controls"></div>
 pconrad:</div>
<blockquote>
<p>I am writing scripts to analyze commit behavior on student software engineering projects. I need to get a list of the files changed so that we can exclude certain commits that only touch documentation and not code.</p>
</blockquote>
</aside>
<p>In addition to using the GitHub API to obtain this information, it might also be interesting to use Git directly. One approach is cloning the repository and periodically fetching those new commits and running a script that automatically compute the list of files changed per commit per author using a command like <a href="https://git-scm.com/docs/git-diff-tree"><code>git diff-tree</code></a> or <a href="https://git-scm.com/docs/git-show"><code>git show</code></a> (passing along the appropriate options).</p>

**Reply by benreaves** [reply]
Cloning all repos and fetching all versions would be very inefficient with time and bandwidth but maybe avoid rate limit? Github must have a rate limit on the number of clones per hour, but I don't see an explicit limit

**Reply by moritzschmitz-oviva** [reply]
You can fetch the base commit first, then query the head commit and its history by filtering `since` and `path`.

**Comment by est** [comment]
I think I hacked a way out.

Suppose you want to check how files under `src` were changed by who's commit

In this example, I check the latest 12 commits, for a depth of two folders max

```
{
  repository(owner: "my", name: "sweet_repo") {
    defaultBranchRef {
      target {
        ... on Commit {
          history(first: 12,path:"src/") {
            nodes {
              message
              oid
              file(path:"src/") {
		object{
                  ... on Tree{
                    entries{
                      path
                      object {
                        ... on Tree{
                          commitUrl
                          entries{
                            path
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```
After the result, parse the `commitUrl` using regex for the `sha` and compare with each `Commit.oid`

If you want to check more path depth, just nest `object{ ... on Tree { entries` like turtles all the way down. All glory to the useful invention of GraphQL!

It doesn't solve all problems but hey it work for me, and it saves tons of stupid N+1 requests.
